### ISSUE TITLE
request: an object to print 'anythings' and maybe also generate them

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1325]
ideally, we could have an 'anything' box, to generate lists or whatever.

But at least a helper abstraction like pddp's [print] or ELSE's [display] is good for Pd's documentation, as discussed in https://github.com/pure-data/pure-data/issues/1320 and also https://lists.puredata.info/pipermail/pd-list/2021-05/129627.html

**Comment by umlaeute** [comment]
nitpick, but we already have an object that *print*s anything. it's aptly called `[print]`.

as for generating 'anythings':
- we already have `[   (` message boxes, but i guess you are talking about something that lets you enter the message in run-mode (rather than edit mode)
- the more important question is: why do you need such a thing?
   - selectors are tightly coupled to the idea of method-calls. changing selectors is really re-programming the patch. but if you want to re-program your patch, use *edit* mode.
   - the only time you might actually want to quickly change selectors is probably in help-patches, demonstrating the different methods. but i don't think this warrants a separate gui-object.
 - there might be use for a "list-gatom" (but you are talking explicitely about 'anything', so i guess this is not what you request)

**Comment by emviveros** [comment]
I don't know if it justify the creation of new object, but the option of a [else/display] and [else/display~] like is very useful in specific case of create documentation, and in debugging patches it void to map multiple endpoints visually in editmode without need to find the pd console windows. Which is very annoying when you have multiple subpatches windows opened in all the area of your monitor.

The use of `[ (` message boxes requires an extra patching to visualize 'anythings' and the space required often makes this alternative unviable to use in the patching process.

**Comment by porres** [comment]
> i guess you are talking about something that lets you enter the message in run-mode (rather than edit mode)

yup

> the only time you might actually want to quickly change selectors

I don't get it, I don't really want to change any selector

> there might be use for a "list-gatom" (but you are talking explicitely about 'anything', so i guess this is not what you request)

there could be both, but my idea of an 'anything' is that you can make it whatever you want and it'd be much like a message box, so you can have a list if you put the 'list' selector. You can have a symbol, etc...

**Comment by umlaeute** [comment]
now i don't get this.

you write:

>  I don't get it, I don't really want to change any selector

and in the same comment you also write:

> so you can have a list if you put the 'list' selector. 

so how do you not want to change the selector and at the same time put a(ny) selector somewhere?

**Comment by umlaeute** [comment]
> is very useful in specific case of create documentation

i think this is a too specific use-case to warrant a special "official" object.

> without need to find the pd console windows. Which is very annoying

not so much.
the Pd-console is just the place where you need to look at to see debugging printout (i also find it very annoying that i need to find and untangle my headphones if i want to hear what Pd makes; the point being that i don't think this is a valid argument).

anyhow: what I do find annoying about the Pd-console is that there is no (visual) connection between a printout and the part of the patch that created that printout.

**Comment by emviveros** [comment]
> anyhow: what I do find annoying about the Pd-console is that there is no (visual) connection between a printout and the part of the patch that created that printout

Totally agree.. Maibe is because that I find so good to use feedback inside patch. But I will point a con which is sequencial order debug is impossible without using [print]

**Comment by porres** [comment]
> so how do you not want to change the selector and at the same time put a(ny) selector somewhere?

By 'putting' a list selector I mean explicitly defining the selector yourself.

The object itself does not create or convert to any selector. It's like a message box, whatever you type in it's what comes out.

exemples:
"boom" - 'anything'
"symbol boom" - a symbol
"cis boom bah" - 'anything'
"list cis boom bah" - a list
"1" -  a float
"1 2 3" - a list
"bang" - a bang
"symbol bang" - a symbol

Above we have the messages and their 'type'. What seems to be confusing is what I mean by 'anything', which I reckon is confusing and not an oficial concept in Pd.

I see it in two ways: 'anything' can be 'any message in Pd' and not just a symbol, a float, a list,  a bang. It can be those, but also other stuff (anything else than this).
But whenever we 'anything else', I also call this message an 'anything type'.

Anyway, I hope I've cleared out the confusion.

For a 'list box', it'd always create the 'list' selector, and if you have just one element, it's create either a float or a symbol, and that's it, you cannot create anything else.

That could be useful, but an 'anything GUI box' would be more versatile and more closely related to a message box. Another thing is that it would display any kind of message it receives (like a [print] does).

So I think this object could be useful in different ways.

> i think this is a too specific use-case to warrant a special "official" object.

An anything message GUI box would cover more ground and use cases. As for something just for a better visual cue on help patches. That is, something more limited like something you cannot type in, then it's less powerful/useful, and it's like pddp/print or else/display (and can be made as an abstraction like these are).

By the way, I use [display] all the time for debugging and other stuff, so it's really just for 'documentation', but I agree that such an abstraction is maybe not that necessary as an 'official' object (as part of 'extra') and I already mentioned that on the Pd list (but I wouldn't oppose putting it in 'extra'). 

But it could be at least a helper abstraction just for the help files, like there arē others (such as [output~]).

**Comment by umlaeute** [comment]
> Anyway, I hope I've cleared out the confusion.

you described quite exactly what I described as "quickly chang[ing] selectors" via a GUI-object. this is (as i've written above) re-programming the patch. i don't think this is a good idea (and of course, you can simply turn your favourite list-gatom into such abomination by addinga `[list trim]`)

**Comment by porres** [comment]
> you can simply turn your favourite list-gatom into such abomination by adding a [list trim])

I can easily trim the list selector with [list trim], I know, but then, whatever I type in there will lose its list selector (or 'symbol' selector if it's just one element). And now if I also want the object to output a 'list message', what do I do? I can't, it's either one or another, but not both. Hence, my idea that you can specify the list selector as you wish (as you do with a message box) is more versatile and it doesn't need a hack that will limit its functionality.

Moreover, a list-gatom would only receive and display list messages. So, conversely, we'd need to convert it to a list before getting displayed and we can't know wether the original message was a list or not...

And what about bangs? Will it deal with bangs? No...

now, why is it an abomination? Pd usually deals with these kind of messages - something like 'open bell.aiff' for [readsf~], for instance, or 'tempo 2 msec' in a [metro]

**Comment by umlaeute** [comment]
@porres wrote:
> but then, whatever I type in there will lose its list selector (or 'symbol' selector if it's just one element). And now if I also want the object to output a 'list message', what do I do?

you type in `list foo bar` which will come out as `[list list foo bar(`, which you convert to `[list foo bar(` by sending through a `[list trim]` object.
if you type `bang` it will come out as `[list bang(`, which gets trimmed to `[bang(`.

for what it is worth, you can already build your desired list-gatom with a symbol-gatom like this:

~~~
|symbol\
|
[list fromsymbol]
|
[fudiparse]
|
~~~

> now, why is it an abomination? Pd usually deals with these kind of messages - something like 'open bell.aiff' for [readsf~], for instance, or 'tempo 2 msec' in a [metro]

yes. this is what i call *code*. if you want to change code, you should switch to edit-mode (which is for code-*editing*).

**Comment by umlaeute** [comment]
@emviveros wrote:
> But I will point a con which is sequencial order debug is impossible without using [print]

how so?
i use `[print]` routinely for debugging sequential ordering.
like always if order matters, the trick is to use `[trigger]`

**Comment by umlaeute** [comment]
both of your examples do not help at when it comes to *sequential order debugging*, which i think was the point why you brought it up here.
actually, **both** your examples are *actively broken* when it comes to debugging data-flow sequential ordering (the left example has the wrong order of output; the right example has *no* order of output).

you have convinced me that `[print]` is superiour for this task, as it at least allows you to get the order right (while `[display]` does not).

 like **always** if order matters, the trick is to use `[trigger]`

**Comment by porres** [comment]
> for what it is worth, you can already build your desired list-gatom with a symbol-gatom like this:

yeah, I had thought about that, but that's not quite as you'll keep getting '\ ' every time you have a space, which is supposed to be always if you have a list and visually confusing, because you don't have a list display, instead you have the display of a symbol that was converted from a list. I know that it is a hack though that kinda does the trick, if you know your way around Pd to build clever hacks. But my point is to offer a more elegant, out of the box solution.

> and an object to display arbitrary messages like this:

Same issues, you have a 'symbol\ oops' instead of 'symbol oops' and an input number '3' comes out as 'float\ 3' in this case too (unlike in [print]).

So yeah, hacks that do not nail it 100%, but I have at least my [else/display] abstraction hack that seems to do the trick perfectly. I also have [else/messbox] which is a compiled GUI sort of in the direction of what I'm proposing here (actually, it'd be a sort of merge between 'display' and 'messbox').

> you type in list foo bar which will come out as [list list foo bar(, which you 
> convert to [list foo bar( by sending through a [list trim] object.
> if you type bang it will come out as [list bang(, which gets trimmed to [bang(.

yeah sure, I missed that, sorry.

So yeah, you see some advantage in having a list-gatom that can be adapted to an 'anything-atom'. I still can't see why though.

> yes. this is what i call code. if you want to change code, you should switch to edit-mode (which is for code-editing).

The answer seems to be here, where you don't consider these as 'messages', but 'code'. This is a new concept to me and my main problem with it is that nowhere in Pd's documentation we see that these are not regular messages, but 'code'.

Anyway, I'd still be quite happy if a list-gatom is what we get, it'd make life much easier in many cases

**Comment by emviveros** [comment]
> you have convinced me that `[print]` is superiour for this task, as it at least allows you to get the order right (while `[display]` does not).

In terms of track multiple bangs (example 1) yes, [print] is superior.

**Comment by emviveros** [comment]
I would just like to reinforce that the inclusion of an object such as [else / display] and [else / display ~] is justified taking into account the criterion of interactivity within the canvas. Even though limited as mentioned in the case of the comparison with the [print], it offers the facility to quickly debug the data type arriving without having to take your eyes off the patch, which is undoubtedly a huge help for creating documentation and collaterally provides an interesting debugging tool, since in many cases the use of [print] causes, besides having to look for the pd console screen, the need to clean it in order to be able to find the information in the console.

For me personally I'm interested in this feature for documentation creation. To patch I'm very happy in use ELSE.

**Comment by porres** [comment]
Something that fulfills the purpose of [else/display~] is less of a hassle to build in pd, it just takes [metro], [snapshot~] and a number box.

The magic glass plugin would also be a great thing for debugging right in the patch.

But let's stick to the focus of this issue request, which is broad enough.

1) An object that displays on the patch whatever message it receives
2) Optionally it could also generate messages as in a new GUI.

Just so we do not get carried away in parallel discussions, let me be objective in what we achieved with this discussion so far. The options:

A) Providing a vanilla abstraction like [else/display] as a helper abstraction in the documentation or as another 'extra' object (this would not be able to generate messages and it'd be more of a documentation helper abstraction).
B) A new GUI like a 'list-box' or 'anything-box' that we can also use for this purpose. This could be a counterpart to the atom boxes or part of iemguis.

We've debated a lot wether an 'anything-box' is an aberration or not, and also that a 'list-box' could be used for 'anything'.

Well, I'm happy either way. And maybe we can even have the two options. Like, a helper abstraction, and a list or whatever box. I dunno...

What I know is that what's easiest on my hands and what I think is the best I can suggest. I can get a full vanilla [display] like helper abstraction (not in 'extra') and I wonder if there's any objection to this. I could then eventually get a PR ready for it.

Creating a GUI box is gonna be real hard for me but if that ever happens later (and I can add a new request just for that even after getting a the helper abstraction in) we can dismiss or delete the helper abstraction (if that made it in first). Something we can also see later is if it's worth getting the helper abstraction into 'extra' (along with 'output~').

So, doe sit sound like a plan? I get a PR with a [display] helper abstraction ready soon and we see about the rest later? Or if @umlaeute or anyone else says they're gonna go for a 'list-box' or something right away, I can hold on to it.

Cheers

**Comment by umlaeute** [comment]
go ahead with your PR (in the end its miller who's going to decide whether it should be included or not).

i don't think i'm going to go for a list-gatom, now that i've seen how simple it is to do that with a symbol-gatom anyhow (and the extra `\` you see before every space is just a bug in the symbol-gatom which is already tracked in #824)

**Comment by porres** [comment]
well, it'd be good to hear your thoughts anyway

and yeah, if we take care of escaping in symbols, and if we make atom boxes flash, I could make a much simpler helper abstraction for the docs too.

**Comment by porres** [comment]
letting this go now that we have a 'list box'
